KAFKA-18635: reenable the unclean shutdown detection

CalvinLiu7947 opened the pull request:
https://issues.apache.org/jira/browse/KAFKA-18635

CalvinLiu7947 committed:
Add Integration tests for the the KIP-966

CalvinLiu7947 committed:
Address comments

CalvinLiu7947 committed:
KAFKA-15586: reenable the unclean shutdown detection

CalvinLiu7947 committed:
fix style

CalvinLiu7947 committed:
temp

CalvinLiu7947 committed:
Minor

CalvinLiu7947 committed:
Fix style

CalvinLiu7947 committed:
Bug fix

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -432,23 +443,29 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
             registerBrokerRecordVersion()));
 
-        if (!request.incarnationId().equals(prevIncarnationId)) {
+        if (isUncleanShutdownRegistration) {
```
Hmm, is this change correct? We should clear the heartbeat as long as the incarnationId is different, right?

CalvinLiu7947 replied:
I think this is debatable. I may miss something. If the heartbeat is only to serve for the replication safety, we may still keep the session for this broker.

junrao replied:
The heartbeat is to keep track of the liveness of a broker since its registration. If the registration changes, we need to reset the heartbeat, right?

CalvinLiu7947 replied:
Refactored the code.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -422,7 +425,15 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
+        }
+        boolean isBrokerStartUp = !request.incarnationId().equals(prevIncarnationId);
+
+        if (isUncleanShutdownRegistration) {
```
Hmm, is this correct? If the incarnationId is different, we always need to generate new PartitionChangeRecords by removing the brokerId from ISR. It seems that we are skipping that step if the previous shutdown is clean in ELR mode.

CalvinLiu7947 replied:
Good question. I think the same principle applies to remove the broker ID from ISR or ELR. The controller only removes the broker from ISR and ELR from the replication/data safety perspective. If the broker is able to prove its data safety(with the same broker epoch read from the disk before its shutdown), the controller should not remove it from ISR and ELR.

splett2 replied:
i don't think that's right. even if the broker shut down and restarted without data loss, we should perform ISR state/leadership transitions.

i think the only behavior difference should be whether or not we remove the broker from the ELR?

CalvinLiu7947 replied:
Yeah, I agree keeping the leader and ISR unchanged may introduce some unexpected impact. Let me update.

junrao reviewed (commented):
@CalvinConfluent : Thanks for the PR. Left a couple of comments.

CalvinLiu7947 committed:
Simplify the code path

CalvinLiu7947 committed:
Merge branch 'trunk' into KAFKA-15582-bug-fix

CalvinLiu7947 committed:
Minor

CalvinLiu7947 committed:
Minor

CalvinLiu7947 commented:
@junrao @splett2 
Based on the feedback, I updated the following:
1. Instead of modifying the whole ClusterControl register logic, I added a clean shutdown flag for the ReplicationControl to update the ISR and ELR accordingly.

- If it is an unclean shutdown, the broker will be removed from ISR and ELR. 

- Otherwise, the broker will be removed from ISR. An election can happen if the broker is the leader. Also, the broker can be added to ELR when possible.

2. I included the integration tests for ELR in this PR.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java:
```diff
@@ -1461,20 +1461,21 @@ void handleBrokerInControlledShutdown(int brokerId, long brokerEpoch, List<ApiMe
      *
-     * @param brokerId      The broker id.
-     * @param records       The record list to append to.
+     * @param brokerId           The broker id.
```
The broker id => The broker id to be shut down

junrao commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -335,7 +335,8 @@ Map<Integer, BrokerRegistration> brokerRegistrations() {
         long newBrokerEpoch,
-        FinalizedControllerFeatures finalizedFeatures
+        FinalizedControllerFeatures finalizedFeatures,
+        boolean uncleanShutdownDetectionEnabled
```
uncleanShutdownDetectionEnabled => cleanShutdownDetectionEnabled ?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -855,13 +866,65 @@ public void testReRegistrationAndBrokerEpoch(boolean newIncarnationId) {
 
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    public void testReRegistrationWithUncleanShutdownDetection(boolean isCleanShutdown) {
```
testReRegistrationWithUncleanShutdownDetection => testReRegistrationWithCleanShutdownDetection ?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -855,13 +866,65 @@ public void testReRegistrationAndBrokerEpoch(boolean newIncarnationId) {
+            clusterControl.brokerRegistrations().get(1).incarnationId());
+        assertFalse(clusterControl.brokerRegistrations().get(1).inControlledShutdown());
+        if (isCleanShutdown) {
+            assertEquals(100, clusterControl.brokerRegistrations().get(1).epoch());
```
This seems incorrect. Independent of whether the shutdown was clean or not, the new broker epoch should always be reflected after broker registration.

CalvinLiu7947 replied:
Good catch. The test is not correct.

junrao commented on metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:
```diff
@@ -460,39 +460,40 @@ public void testUncleanShutdownBrokerElrEnabled() throws Throwable {
 
             // Unclean shutdown should not remove the last known ELR members.
-            active.registerBroker(
+            reply = active.registerBroker(
```
reply is unused.

junrao commented on core/src/test/java/kafka/server/integration/EligibleLeaderReplicasIntegrationTest.java:
```diff
@@ -0,0 +1,450 @@
+        Producer producer = null;
+        Consumer consumer = null;
+        try {
+            // check which partition is on broker 0 which we'll kill
```
There is only 1 partition. Ditto below.

CalvinLiu7947 replied:
Looks like this comment is not useful. Removed.

junrao commented on core/src/test/java/kafka/server/integration/EligibleLeaderReplicasIntegrationTest.java:
```diff
@@ -0,0 +1,450 @@
+            consumer.subscribe(Collections.singleton(testTopicName));
+
+            producer.send(new ProducerRecord<>(testTopicName, "0", "0")).get();
+            Thread.sleep(1000);
```
Hmm, why do we need to sleep here? Could we use `waitUntil` ?

junrao commented on core/src/test/java/kafka/server/integration/EligibleLeaderReplicasIntegrationTest.java:
```diff
@@ -0,0 +1,450 @@
+
+            // Now the partition is under min ISR. HWM should not advance.
+            producer.send(new ProducerRecord<>(testTopicName, "1", "1")).get();
+            Thread.sleep(1000);
```
1 sec is quite long. How about 100ms?

junrao commented on core/src/test/java/kafka/server/integration/EligibleLeaderReplicasIntegrationTest.java:
```diff
@@ -0,0 +1,450 @@
+
+            topicPartitionInfo = adminClient.describeTopics(Collections.singletonList(testTopicName))
+                .allTopicNames().get().get(testTopicName).partitions().get(0);
+            assertEquals(2, topicPartitionInfo.elr().size());
```
This seems unnecessary since it's covered by the `waitForIsrAndElr` call earlier?

junrao commented on core/src/test/java/kafka/server/integration/EligibleLeaderReplicasIntegrationTest.java:
```diff
@@ -0,0 +1,450 @@
+                .allTopicNames().get().get(testTopicName).partitions().get(0);
+            int lastKnownLeader = topicPartitionInfo.lastKnownElr().get(0).id();
+
+            brokers().foreach(broker -> {
```
One of the brokers is not killed. We should avoid deleting its file, right?

CalvinLiu7947 replied:
This test verifies the behaviors of unclean shutdown broker register.  We first bring down all 4 brokers. Then, the first 3 brokers coming up verify that unclean shutdown will remove the broker from ELR. The last broker verifies that un unclean shutdown broker can be the leader because it is the lastKnownLeader.

junrao replied:
Hmm, in `generateConfigs()`, numConfigs is 5. So, there should be 5 brokers in total?

CalvinLiu7947 replied:
I misunderstood. Yes, we need one additional broker to handle the admin client requests.
Updated the not operate on the last broker.

junrao reviewed (commented):
@CalvinConfluent : Thanks for the updated PR. A few more comments.

CalvinLiu7947 committed:
Address comments

CalvinLiu7947 committed:
Minor

CalvinLiu7947 committed:
Minor

junrao reviewed (commented):
@CalvinConfluent : Thanks for the updated PR. Just one comment.

junrao commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -2203,7 +2204,7 @@ QuorumControllerMetrics controllerMetrics() {
 
-    void handleUncleanBrokerShutdown(int brokerId, List<ApiMessageAndVersion> records) {
-        replicationControl.handleBrokerUncleanShutdown(brokerId, records);
+    void handleBrokerShutdown(int brokerId, boolean isUncleanShutdown, List<ApiMessageAndVersion> records) {
```
isUncleanShutdown => isCleanShutdown ?

CalvinLiu7947 committed:
Address comments

CalvinLiu7947 committed:
Fix test

junrao reviewed (commented):
@CalvinConfluent : Thanks for the updated PR. The code LGTM. Are the flaky test failures related to this PR, especially EligibleLeaderReplicasIntegrationTest?

CalvinLiu7947 committed:
Try fix the integration flaky

CalvinLiu7947 committed:
Fix again

CalvinLiu7947 committed:
Minor

CalvinLiu7947 commented:
@junrao Thanks for the review, now the tests look good.

junrao reviewed (approved):
@CalvinConfluent : Thanks for triaging the tests. LGTM

junrao commented:
@CalvinConfluent : Do you want to create a cherry-pick PR for 4.0 branch just to make sure all tests pass there?

CalvinLiu7947 commented:
@junrao Sure, will do. But it may take a while as there are other required PRs pending merging into 4.0
